src/lightning/data/dataset.py

```diff
@@ -0,0 +1,89 @@
+import os
+import tempfile
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Optional, Tuple
+
+from torch.utils.data import Dataset as TorchDataset
+
+from lightning.data.authenticators import _DatasetBackend, LocalDatasetBackend, S3DatasetBackend
```

Comment:
Maybe call this `lightning.data.backends`?

Reply:
This is where the actual functionality will end up, while Dataset will just expose hooks, so that's more future-proof.